kubernetes/kubernetes
# Fix PodAntiAffinity issues in case of multiple affinityTerms

- rename podMatchesAffinityTermProperties to podMatchesAllAffinityTermProperties
- add podMatchesAnyAffinityTermProperties which should be used in some PodAntiAffinity cases
- add unit tests

**What this PR does / why we need it**:

There are 2 phases to verify symmetry of PodAntiAffinity:
- verify labels of incoming pod not violates antiaffinity of existing pods
- verify antiaffinity of incoming pod not violates labels of existing pods

This PR is to fix issue in second step.

Currently logic of second step tries to invoke `podMatchesAffinityTermProperties` when checking both affinity and antiaffinity. 

It's not that accurate, as `podMatchesAffinityTermProperties` is a "match all" semantics, say for each term, ensure it matches target selector/labels, and return true if all term passes check. This is good for affinity, but not antiaffinity.

Regarding antiaffinity, it's a "match any" semantics - i.e. if found any term match target selector/labels, return true.

So without this fix, unit tests (which is added in this PR) failed on following cases:
- a) incoming pod's anti-affinity has more than one term - existing pod violates all terms
- b) incoming pod's anti-affinity has more than one term - existing pod violates partial term - case 1
- c) incoming pod's anti-affinity has more than one term - existing pod violates partial term - case 2

Case a) is guarded in before-mentioned step 1 of symmetric logic. So we don't see it in a running k8s cluster. Case b) and c) are real problems and specifically case c) is what issue #68101 reported.


**Which issue(s) this PR fixes**:
Fixes #68101

**Special notes for your reviewer**:
After reviewing the correctness of this PR, bsalamat and k82cn can decide if it needs to go into 1.12 release cycle. 

**Release note**:
```release-note
Fix PodAntiAffinity issues in case of multiple affinityTerms.
```
